beluga_amcl/src/ndt_amcl_node_3d.cpp

```diff
@@ -0,0 +1,503 @@
+            execution_policy);
+      },
+      get_motion_model(), get_execution_policy());
+  return std::make_unique<NdtAmclVariant>(std::move(amcl));
+}
+
+void NdtAmclNode3D::do_initial_pose_callback(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr message) {
+  auto pose = Sophus::SE3d{};
+  tf2::convert(message->pose.pose, pose);
+  const Eigen::Map<const Sophus::Matrix6d> covariance(message->pose.covariance.data());
+  last_known_estimate_ = std::make_pair(pose, covariance);
+  last_known_odom_transform_in_map_.reset();
+  initialize_from_estimate(last_known_estimate_.value());
+}
+
+void NdtAmclNode3D::do_periodic_timer_callback() {
+  if (!particle_filter_) {
+    return;
+  }
+
+  if (particle_cloud_pub_->get_subscription_count() == 0) {
+    return;
+  }
+  std::visit(
+      [this](const auto& particle_filter) {
+        auto message = beluga_ros::msg::PoseArray{};
+        beluga_ros::assign_particle_cloud(particle_filter.particles(), message);
+        beluga_ros::stamp_message(get_parameter("global_frame_id").as_string(), now(), message);
+        particle_cloud_pub_->publish(message);
+      },
+      *particle_filter_);
+}
+
+// TODO(alon): Wouldn't it be better in the callback of each message to simply receive
+// it and define another timer or thread to do the work of calculation and publication?
+void NdtAmclNode3D::laser_callback(sensor_msgs::msg::PointCloud2::ConstSharedPtr laser_scan) {
+  if (!particle_filter_ || !last_known_estimate_.has_value()) {
+    RCLCPP_WARN_THROTTLE(
+        get_logger(), *get_clock(), 2000, "Ignoring laser data because the particle filter has not been initialized");
+    return;
+  }
+
+  auto base_pose_in_odom = Sophus::SE3d{};
+  auto laser_pose_in_base = Sophus::SE3d{};
+  try {
+    // Use the lookupTransform overload with no timeout since we're not using a dedicated
+    // tf thread. The message filter we are using avoids the need for it.
+    tf2::convert(
+        tf_buffer_
+            ->lookupTransform(
+                get_parameter("odom_frame_id").as_string(), get_parameter("base_frame_id").as_string(),
+                tf2_ros::fromMsg(laser_scan->header.stamp))
+            .transform,
+        base_pose_in_odom);
+    tf2::convert(
+        tf_buffer_
+            ->lookupTransform(
+                get_parameter("base_frame_id").as_string(), laser_scan->header.frame_id,
+                tf2_ros::fromMsg(laser_scan->header.stamp))
+            .transform,
+        laser_pose_in_base);
+  } catch (const tf2::TransformException& error) {
+    RCLCPP_ERROR(get_logger(), "Couldn't find needed transform, : %s", error.what());
+    return;
+  }
+
+  const auto update_start_time = std::chrono::high_resolution_clock::now();
+
+  pcl::PointCloud<pcl::PointXYZ> pcl;
+  std::vector<Eigen::Vector3d> measurement;
+  pcl::fromROSMsg(*laser_scan, pcl);
+  measurement.reserve(pcl.points.size());
+
+  // TODO (serraramiro1): Can this be more efficient?
```

Comment:
@serraramiro1 well, we could use a [`PointCloud2Iterator`](https://github.com/ros2/common_interfaces/blob/rolling/sensor_msgs/include/sensor_msgs/point_cloud2_iterator.hpp) for each coordinate field, then zip them, then transform them to `Eigen::Vector3d` and apply the transform one by one. We still have to apply the transform for each point, but at least we won't be bringing PCL as a dependency.

Reply:
Another intrusive thought. Point clouds will always be super dense, but NDT maps are sparse. Would it more efficient to transform the map instead? Food for thought, wouldn't try that here.

Reply:
Good point, although we first apply NDT to the clouds and then we transform, for every particle, so we do this (somewhat costly) transform only once. We could update the measurement type to take in the already NDT transformed type and maybe make this even cheaper